Add method for external storage

natelust opened the pull request:
Android supports accessing the "external" storage medium. Support
finding this path in flutter with a method call in path_provider

I am interested in writing a flutter app that will access some files a user may have put on their device manually (android). This necessitates looking up the path for the external storage. Though I could do the same thing in my own package, I figured it would be more useful to have this functionality in the path provider package so it can be more broadly used. Sadly this functionality is restricted to android only, as iOS does not let you escape the app sandbox. If there is some other mechanism you would prefer me to go through instead of just creating a pull request, please let me know.

natelust commented:
I signed it!

mit-mit commented on packages/path-provider/lib/path_provider.dart:
```diff
@@ -40,3 +40,16 @@ const _channel = const MethodChannel('plugins.flutter.io/path_provider');
+
+  /// Path to a directory where the application may access top level storage
+  ///
+  /// On iOS, This function returns null as it is not possible to access outside
```
Sentences in comments are missing terminating `.`

Few typos in iOS sentence, change to: 
```
/// On iOS, this function returns null, as it is not possible to access the
/// file system outside the apps sandbox.
```

szakarias commented on packages/path-provider/android/src/main/java/io/flutter/plugins/path_provider/PathProviderPlugin.java:
```diff
@@ -46,4 +51,8 @@ private String getPathProviderTemporaryDirectory() {
   }
+
+  private String getPathProviderStorageDirectory() {
+    return Environment.getExternalStorageDirectory().getAbsolutePath();
```
What is the use case for this? The documentation for [getExternalDirectory()](https://developer.android.com/reference/android/os/Environment.html#getExternalStorageDirectory()) says not to use this directly but instead use [getExternalStoragePublicDirectory(String)](https://developer.android.com/reference/android/os/Environment.html#getExternalStoragePublicDirectory(java.lang.String)) .

natelust replied:
The use case I had in mind for this was to get the root of the available file system, such that a file browser could be created within an app. I wanted to create a toy audio book app that would allow the user to navigate to a directory and mark it as the location all audio books could be found under. The latter API call you bring up is not quite targeted for this use case, as it expects you to specify what type of media you are interested in (i.e. for finding the location to save pictures). In my case the user may have located their audio books at any place accessible on the external storage.

szakarias commented on packages/path-provider/lib/path_provider.dart:
```diff
@@ -40,3 +40,16 @@ const _channel = const MethodChannel('plugins.flutter.io/path_provider');
+  /// the apps sandbox
+  ///
+  /// On Android this returns getExternalStorageDirectory
+  Future<Directory> getExternalStorageDirectory() async {
```
On iOS, this will result in a [MissingPluginException](https://docs.flutter.io/flutter/services/MissingPluginException-class.html) since there is no handling of  a method with that name. To avoid this, you could [check the OS](https://docs.flutter.io/flutter/dart-io/Platform-class.html) here and throw for instance a [StateError](https://www.dartdocs.org/documentation/flutter/0.0.21/dart-core/StateError-class.html). Remember to document also, that the OS needs to be determined before calling this method.

natelust replied:
I added a check for iOS and documented in the comments above the function

natelust commented:
I have added to the example to make use of this new function, but I unfortunately could not test the example, as it seems to be broken by a change in flutter itself (most likely due to moving this stuff out to stand alone). I followed the structure of the existing example, so I don't expect any issues. I would be happy to look into fixing the example as a new small project.

I appreciate you considering this pull request. I am quite new to mobile app development. (I am coming from a background of python/c++ for high performance computing in astrophysics). Flutter is the first framework that has gotten my genuinely interested in the mobile ecosystem. I really like the design choices and am enjoying Dart as a language. I appreciate the opportunity to contribute to this project.

natelust commented:
Also, I don't know the etiquette on this project in regards to cleaning up commits, and commit history in general. Is it advisable to squash the commits down like I did?

szakarias commented on packages/path-provider/example/lib/main.dart:
```diff
@@ -105,6 +113,24 @@ class _MyHomePageState extends State<MyHomePage> {
               child: new FutureBuilder<Directory>(
                   future: _appDocumentsDirectory, builder: _buildDirectory),
             ),
+            new Cloumn(
```
Cloumn -> Column

szakarias commented on packages/path-provider/example/lib/main.dart:
```diff
@@ -105,6 +113,24 @@ class _MyHomePageState extends State<MyHomePage> {
+            new Cloumn(
+              children : <Widget>[
+                new Padding(
+                  padding: const EdgeInserts.add(16.0),
```
EdgeInserts.add(16.0) -> EdgeInsets.all(16.0)

szakarias commented on packages/path-provider/example/lib/main.dart:
```diff
@@ -105,6 +113,24 @@ class _MyHomePageState extends State<MyHomePage> {
+                new Padding(
+                  padding: const EdgeInserts.add(16.0),
+                  child: new RaisedButton(
+                    child: const Test('${Platform.operatingSystem != "ios" ?
```
const Test -> New Text 

You cannot have const here as the expression is not constant.

szakarias commented on packages/path-provider/example/lib/main.dart:
```diff
@@ -105,6 +113,24 @@ class _MyHomePageState extends State<MyHomePage> {
+                  child: new RaisedButton(
+                    child: const Test('${Platform.operatingSystem != "ios" ?
+                                      "Get External Storage Directory" :
+                                      "External Directories are unavailable " +
```
Directories -> directories

szakarias commented on packages/path-provider/example/lib/main.dart:
```diff
@@ -105,6 +113,24 @@ class _MyHomePageState extends State<MyHomePage> {
+                                      "Get External Storage Directory" :
+                                      "External Directories are unavailable " +
+                                      "on iOS"}'),
+                    onPressed: _requestExternalStorageDirectory,
```
Let `onPressed` be null on iOS. This disables the button.

szakarias commented on packages/path-provider/example/lib/main.dart:
```diff
@@ -65,6 +66,13 @@ class _MyHomePageState extends State<MyHomePage> {
 
+  void _requestExternalStorageDirectory() {
+    setState(() {
+      if (Platform.operatingSystem != "ios")
```
Please check the OS with `Platform.isIOS` instead, here and below as well.

szakarias commented on packages/path-provider/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## [0.2.1] - 2017-05-11
```
Once we reach the final version this date should be updated accordingly.

szakarias commented on packages/path-provider/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## [0.2.1] - 2017-05-11
+
+* Add function to determine external storage directory
```
Missing terminating `.`

szakarias commented on packages/path-provider/lib/path_provider.dart:
```diff
@@ -40,3 +40,20 @@ const _channel = const MethodChannel('plugins.flutter.io/path_provider');
+  /// On Android this returns getExternalStorageDirectory.
+  Future<Directory> getExternalStorageDirectory() async {
+    if (Platform.operatingSystem == "ios")
+      throw new UnimplementedError("Functionality not available on iOS");
```
Please use `UnsupportedError` instead.

natelust commented:
I have addressed the issues mentioned above (except updating the final date), and rebased my fork onto master. I must say I am actually quite embarrassed at some of the mistakes I had. Previously I had assumed the build failures I was seeing in the platform specific libraries were related to API changes, but it was my own silly bugs causing it to fail. I must spend more time looking over things in the future. You have been quite patient in working with me, and I thank you for that. I am sorry that this has caused so much work for you. I have two commits related to the changes requested, which can be squashed prior to merging in whatever manor you prefer. Please let me know if there is anything else to be done.

szakarias commented on packages/path-provider/android/src/main/java/io/flutter/plugins/path_provider/PathProviderPlugin.java:
```diff
@@ -5,13 +5,16 @@
 import io.flutter.util.PathUtils;
 
+
+
```
Remove empty lines

natelust replied:
Should I squash down to one commit?

szakarias replied:
You can use the 'Squash and merge' button, and remove the extra commit messages.

natelust replied:
I have no write access to the github repository, so I can not actually do the merge myself. I have squashed all the commits on my fork, so it can be merged at any point.

mit-mit commented:
Yup, we will handle the merge. Thanks so much for contributing to Flutter, @natelust !

badaween-ahmed commented:
Any update on adding support for external Storage (like Sd Card, External Drive)?
